AI/Nullkiller/AIGateway.cpp

```diff
@@ -488,6 +489,24 @@ void AIGateway::showWorldViewEx(const std::vector<ObjectPosInfo> & objectPositio
 	NET_EVENT_HANDLER;
 }
 
+boost::optional<BattleAction> AIGateway::makeSurrenderRetreatDecision(
+	const BattleStateInfoForRetreat & battleState)
+{
+	LOG_TRACE(logAi);
+	NET_EVENT_HANDLER;
+
+	double fightRatio = battleState.getOurStrength() / (double)battleState.getEnemyStrength();
+
+	// if we have no towns - things are already bad, so retreat is not an option.
+	if(cb->getTownsInfo().size() && fightRatio < 0.3 && battleState.canFlee)
```

Comment:
1) 0.3 is a magic const
2) if enemy hero has a 1 fast unit, he still can throw magic arrow. I suggest do make heuristic smarter and check if AI make magic action before surrender

Reply:
Spellcast is evaluated before retreat so I expect AI should try it first. 

2 - Smarter things are always welcome! You can check the PR from description. It went exactly this path and actually is smarter than mine. It can detect when it is last turn before hero dies. But for whatever reason it was not merged. Guess why? I can - to many things at once. More code - more bugs - more time to fix - higher chance for PR to be abandoned.
AI code is never ideal, always there is a room for improvement, always in some cases decisions will be not optimal. If you try to fix all it at once - it will just never end.

Reply:
I agree that approach "better than now" is what we should target for. 
But are your sure this behavior is not possible?
1) enemy hero attacks player hero
2) enemy hero immediately retreats

Reply:
For NKAI it is not allowed to attack if power ratio is below safe_attack constant which is higher than 1 as far as I remember.
Also AI prefer attacking when his power is actually higher

Reply:
I will fix magic constant in battle ai PR